### PR TITLE
Replace the video in the triangles intro with a shader animation

### LINKED ISSUE #5805: Replace triangle intro sequence video with a shader (or otherwise) implementation
I have a WQHD Monitor with is 2560 × 1440 pixels and the video with a resolution of 1080 × 1080 pixels looks pretty pixelated since it starts at being scaled up to 120% and scales down to 80%.
I suggest a resolution of 1,296 × 1,296 pixels for SD and double resolution for HD and up.

https://github.com/ppy/osu-resources/blob/master/osu.Game.Resources/Textures/Menu/logo-triangles.mp4

https://github.com/ppy/osu/blob/76d34967a4ba0acd3a510c17921c2c43da194816/osu.Game/Screens/Menu/IntroTriangles.cs#L221

### PR DESCRIPTION
**Pull request by jorolf**
- Closes #5805
- [x] Depends on https://github.com/ppy/osu-resources/pull/105

I've gone ahead and forked @Wieku 's branch since he hasn't worked on it for 6 months.

The animation class has been simplified and more generalized. (Although I doubt it's going to be used somewhere again)

The only thing that's broken right now are the animation speeds and colours as I don't have Adobe After Effects.

### CODE REVIEW

File: osu.Game/Screens/Menu/IntroTriangles.cs [hunk 1]
```diff
@@ -259,20 +265,44 @@ protected override void LoadComplete()
 
             private class LazerLogo : CompositeDrawable
             {
-                private readonly Stream videoStream;
+                private HueAnimation highlight, animation;
```
Comment (bdach) [inline osu.Game/Screens/Menu/IntroTriangles.cs:268]: nit: would probably rename to something like `outline` and `outlineHighlight`

**Review by bdach — COMMENTED** [review]
Seems to work ok on my linux box. The asset in resources looks visibly off due to compression effects, however - we should see if we can get it to a workable state first.

File: osu.Game/Graphics/Sprites/HueAnimation.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        }
+
+        public override bool IsPresent => true;
+
+        protected override DrawNode CreateDrawNode() => new HueAnimationDrawNode(this);
+
+        private class HueAnimationDrawNode : SpriteDrawNode
+        {
+            protected new HueAnimation Source => (HueAnimation)base.Source;
```
[the remark this answers is not in the document]
Comment (jorolf) [reply osu.Game/Graphics/Sprites/HueAnimation.cs:43]: Going to fix this, I just copied this from other drawnodes

**Comment by peppy** [comment]
Before we go any further, the animation needs to match. Currently it doesn't match at all (the curves are wrong and the gradient is wrong – the colours don't get dimmer after the initial lighting as they do in the video).

No further consideration should even be given to image compression unless this can first be fixed.

**Comment by jorolf** [comment]
Colours can be changed very easily, the shader uses the colour property of drawables

**Comment by peppy** [comment]
The issue is the trail is not fading, not the colours.

**Comment by jorolf** [comment]
I have updated the animation texture to better represent the original animation. However I wasn't quite able to figure out a timing/easing connection between the outline and the highlight meaning that those are still not correct. Maybe I could seperate them into 2 seperate texture and overlay them ontop of each other....

**Comment by peppy** [comment]
Can you explain what you mean by the outline / highlight?

**Comment by jorolf** [comment]
edit

The outline is the gray line and the highlight is the white line.

**Comment by peppy** [comment]
@arflyte are you able to provide the curves for the two pieces in as much detail as possible so we can try to reproduce them?

**Comment by arflyte** [comment]
AfterFX_200702-183631

This was the curve used. Purple is used for End Trim, and Orange is Start Trim. It follows according to the tune of the music.

There's two set of paths: Normal and Highlight. Normal only uses the End Trim. The highlight uses both Start and End Trim.

Here's the AE file, just in case you wanna a closer look at it.
[Lazer Intro folder.zip](https://github.com/ppy/osu/files/4863333/Lazer.Intro.folder.zip)

**Comment by jorolf** [comment]
Do those curves have specific functions?

Alternatively, I created a small program that "compiles" several frames into an animation, however it's kind of hard to differentiate between the highlight and the normal path. I could create an animation without the highlight from a video where the highlight is made the same color as the background (black)

**Comment by arflyte** [comment]
I didn't use any specific curve functions for it, manually tweaked the curves in AE.

**Comment by peppy** [comment]
Are you able to re-export with the white layer changed to black for @jorolf to try and work with?

**Comment by arflyte** [comment]
Exported them lines into two separate videos.

https://r.flyte.moe/2007/Logo%20Lazer_1.mp4
https://r.flyte.moe/2007/Logo%20Lazer.mp4

**Comment by jorolf** [comment]
grafik
I just noticed that there is a very faint background difference between those two points:
grafik
is that intended? (And is my monitor gamma possibly misconfigured? 🤔 I can only see it on one of my two monitors)

**Comment by peppy** [comment]
I'd say that boils down to video compression artifacts and ignore it.

**Comment by peppy** [comment]
Tested on mobile just to ensure the shader is sane, looking good 👍 .

**Review by peppy — DISMISSED** [review]

**Pull request merged**